### ISSUE TITLE
MatrixExpr `as_explicit` cannot work when dimensions are symbolic

### ISSUE THREAD

**Issue by goretkin** [diofant/diofant#469]
```
In [1]: import diofant
                  
In [2]: n = diofant.Symbol("n")

In [3]: A = diofant.MatrixSymbol("A", n, n)

In [4]: B = diofant.MatrixSymbol("B", n, n)

In [5]: diofant.Eq(A, B)
Out[5]: A = B

In [6]: diofant.simplify(diofant.Eq(A, B))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-09f9fbbfdcc3> in <module>()
----> 1 diofant.simplify(diofant.Eq(A, B))

/Users/goretkin/repos/diofant/diofant/simplify/simplify.py in simplify(expr, ratio, measure, fu)
    588 
    589     try:
--> 590         return expr._eval_simplify(ratio=ratio, measure=measure)
    591     except AttributeError:
    592         pass

/Users/goretkin/repos/diofant/diofant/core/relational.py in _eval_simplify(self, ratio, measure)
    181             # approximation.  If .equals(0) gives None, cannot be deduced.
    182             if not dif.has(Dummy, Symbol):
--> 183                 know = dif.equals(0)
    184                 if know:
    185                     dif = S.Zero

/Users/goretkin/repos/diofant/diofant/matrices/expressions/matexpr.py in equals(self, other)
    303         True
    304         """
--> 305         return self.as_explicit().equals(other)
    306 
    307     def canonicalize(self):

/Users/goretkin/repos/diofant/diofant/matrices/expressions/matexpr.py in as_explicit(self)
    259         return ImmutableMatrix([[    self[i, j]
    260                             for j in range(self.cols)]
--> 261                             for i in range(self.rows)])
    262 
    263     def as_mutable(self):

TypeError: 'Symbol' object cannot be interpreted as an integer

In [7]: 
```

I would expect `simplify` to be a no-op, instead of erroring.

**Comment by skirpichev** [comment]
Probably, we should separate case of symbolic shapes in equals() method.

Quick exit for non-numeric matrix shapes - works for me:
```diff
diff --git a/diofant/matrices/expressions/matexpr.py b/diofant/matrices/expressions/matexpr.py
index 1ba9fa7..38994ab 100644
--- a/diofant/matrices/expressions/matexpr.py
+++ b/diofant/matrices/expressions/matexpr.py
@@ -302,6 +302,8 @@ def equals(self, other):
         >>> Identity(3).equals(eye(3))
         True
         """
+        if not all(x.is_number for x in self.shape):
+            return
         return self.as_explicit().equals(other)
 
     def canonicalize(self):
```

**Comment by goretkin** [comment]
That looks like a good solution to me.

**Comment by skirpichev** [comment]
Ok, let me know if you have better ideas.  Meanwhile, I'll include this fix in #467.
